### PR TITLE
Path publisher

### PR DESCRIPTION
**Pull request by ThijsBeurskensTUE**

### CODE REVIEW

File: src/communication.cpp [hunk 1]
```diff
@@ -226,6 +261,29 @@ void Communication::bumperbCallback(const std_msgs::BoolConstPtr& msg)
...
+
+    double roll, pitch, yaw;
+    T.getRPY(roll, pitch, yaw);
+
+    mapconfig.mapOrientation = yaw + M_PI/2;
+
+    mapconfig.mapResolution = msg->resolution;
+
+    mapconfig.mapOffsetX = (msg->width * mapconfig.mapResolution / 2) * cos(yaw)-(msg->height * mapconfig.mapResolution / 2) * sin(yaw) + msg->origin.position.x;
+    mapconfig.mapOffsetY = (msg->width * mapconfig.mapResolution / 2) * sin(yaw)+(msg->height * mapconfig.mapResolution / 2) * cos(yaw) + msg->origin.position.y;
```
Comment (PetervDooren) [inline src/communication.cpp:282]: You encountered an issue with the origin of the world not matching the map. Could this be the cause? This formula might not be correct

File: src/io.cpp [hunk 2]
```diff
@@ -87,4 +93,107 @@ bool IO::ok()
...
+    pathMarker.color.a = 1.0;
+    pathMarker.color.r = color[0];
+    pathMarker.color.g = color[1];
+    pathMarker.color.b = color[2];
+    pathMarker.pose.orientation.w = 1.0;
+    pathMarker.scale.x = width;
+    for (std::vector<std::vector<double>>::iterator it = path.begin(); it != path.end(); ++it)
+    {
+        geometry_msgs::Point point;
+        if ((*it).size() < 2)
```
Comment (PetervDooren) [inline src/io.cpp:116]:
```suggestion
        if (it->size() != 2)
```
Also break if they added more then 2 values. That is also not intended behaviour and probably indicates that something went wrong.

File: src/io.cpp [hunk 3]
```diff
@@ -87,4 +93,107 @@ bool IO::ok()
...
+            mapOffset.setOrigin(tf2::Vector3(mapData.mapOffsetX, mapData.mapOffsetY, 0));
+            tf2::Quaternion q;
+            q.setRPY(0, 0, mapData.mapOrientation);
+            mapOffset.setRotation(q);
+
+            tfPoint = mapOffset * tfPoint;
+            point.x = tfPoint.getOrigin().x();
+            point.y = tfPoint.getOrigin().y();
+            point.z = tfPoint.getOrigin().z();
+        }
```
Comment (PetervDooren) [inline src/io.cpp:151]: What if the mapdata is not initialised? Do you still publish in map frame

File: src/io.cpp [hunk 4]
```diff
@@ -87,4 +93,107 @@ bool IO::ok()
...
+
+        pathMarker.points.push_back(point);
+    }
+    if (pathMarker.points.size() < 2)
+    {
+        ROS_ERROR_STREAM("not enough valid points");
```
Comment (PetervDooren) [inline src/io.cpp:157]: specify how many points were found. That helps debugging. In this case you might also print out these points. e.g.

"not enough valid points. no valid points"
"not enough valid points. only got [x1,y1]"

**Review by PetervDooren — CHANGES_REQUESTED** [review]
#TODO update emc-env to ensure installation of the required packages:

- nav_msgs
- tf2
- tf2_ros

File: src/io.cpp (hunk 3 above)
Comment (PetervDooren) [reply src/io.cpp:151]: I'll rephrase. I think its better to throw an error and return false in this situation.

File: src/io.cpp (hunk 4 above)
Comment (PetervDooren) [reply src/io.cpp:157]: But a single, general check on the size is sufficient.

File: src/communication.cpp (hunk 1 above)
Comment (ThijsBeurskensTUE) [reply src/communication.cpp:282]: This formula correctly finds the centre of the map, but the orgin of the simulator is one pixel off centre. I think it's better to correct the simulator to have the origin in the centre than to introduce this offset everywhere.

**Review by KdVos — APPROVED** [review]

**Pull request merged**